원문: Moler, Cleve B. Numerical computing with MATLAB. Society for Industrial and Applied Mathematics, 2008.

레오나도 피사노 피보나치는 1170년 무렵 태어나 1250년 오늘날의 이탈리아에 속한 피사에서 세상을 떠났다. 그는 유럽과 북아프리카를 두루 여행했었다. 그는 몇권의 수학 문헌을 썼는데 특히 유럽에 인도-아라비아 숫자를 소개했다. 그의 책은 손으로 배껴썼어야 했지만, 널리 읽혀졌다. 그의 가장 잘 알려진 책 *Liber Abaci* 는 1202년에 출판되었는데, 다음과 같은 문제를 제시하였다.

>> *어떤 사람이 한쌍의 토끼를 사방이 벽으로 둘러싸인 장소에 두었다. 그 한쌍으로부터 몇쌍의 토끼가 1년 이내에 태어날 수 있겠는가? 매달 각 쌍이 새로운 한쌍을 낳고, 태어난 후 두달째 부터 새끼를 낳을 수 있다.*

오늘날 이 문제의 해는 *피보나치 수열* 또는 *피보나치 수* 로 알려져 있다. 피보나치 수에 기반하는 자그마한 수학적 산업이 있다. "피보나치"를 인터넷에서 검색해 보면 수십개의 웹사이트와 수백 페이지의 문헌을 찾을 수 있을 것이다.  피보나치 협회도 있어서 *Fibonacci Quarterly*라는 학술지도 출간한다.

피보나치가 새로 태어난 한 쌍이 성장하는데 1달 걸린다는 것을 명시하지 않았다면, 그의 이름을 딴 수열이 존재하지는 않았을 것이다.  매달 단순히 두배로 늘어날 것이다.  *n* 개월 후에는 $2^n$ 쌍의 토끼가 있을 것이다.  토끼의 수는 엄청나게 불어나겠지만, 무언가 수학적으로 독특한 점은 없었을 것이다.

$f_n$ 이 *n* 개월 후의 토끼 쌍의 수라고 하자. 요점은 월말의 토끼의 수는 월초의 토끼의 수에 성장한 쌍이 낳은 수를 합한 것이라는 것이다:

$f_n = f_{n-1} + f_{n-2}$

초기 조건은 첫 달에는 한쌍의 토끼가 있고, 둘째 달에는 두쌍이 있다는 것이다.

$f_1 = 1, f_2 = 2$


In [1]:
import numpy as np


def fibonacci(n):
    """
    Fibonacci sequence
    generates the first n Fibonacci numbers
    :param n: 
    :return: 
    """
    
    f = np.zeros((n, 1))
    f[0] = 1
    f[1] = 2
    for k in range(2, n):
        f[k] = f[k - 1] + f[k - 2]
    
    return f


fibonacci(12)

array([[   1.],
       [   2.],
       [   3.],
       [   5.],
       [   8.],
       [  13.],
       [  21.],
       [  34.],
       [  55.],
       [  89.],
       [ 144.],
       [ 233.]])

답은 233 쌍이라는 것이다. (만일 매달 두배로 12달 동안 늘어났다면 4096 쌍이 될 것이다)

함수를 자세히 살펴보자. 첫 행은 

``import numpy as np``

으로 ``numpy`` 라는 기능 module 을 ``np`` 라는 이름으로 불러 들인다. 그 다음 행은

``def fibonacci(n):``

이다.  첫번째 단어는 이것이 함수 선언의 시작임을 알린다. 나머지 부분은 함수의 이름과 이 함수가 한개의 입력 매개변수 ``n`` 을 받아들인다는 점을 표시한다. 그 다음에는 ``help`` 에 표시되는 내용을 담고 있다.

In [2]:
help(fibonacci)

Help on function fibonacci in module __main__:

fibonacci(n)
    Fibonacci sequence
    generates the first n Fibonacci numbers
    :param n: 
    :return:



그 다음 행

``f = np.zeros((n, 1))``

은 0.0으로 채워진 $n \times 1$ 배열을 만들어 ``f`` 에 할당한다.  한 열로 이루어진 배열은 열 벡터, 한 행으로 이루어진 배열은 행 벡터로 생각될 수 있다. (``numpy.array`` 와 ``numpy.matrix`` 는 조금 다르다.)

다음 두 행

``f[0] = 1``

``f[1] = 2``

는 초기 조건을 제공한다.

그 다음은 ``for``문으로 모든 연산을 실시한다.

``    for k in range(2, n):``

>``        f[k] = f[k - 1] + f[k - 2]``

``for`` 문과 ``if`` 문에는 ``:`` 기호와 들여쓰기가 적용된다.

마지막 행은 반환값을 ``f``로 지정한다.

이 예제는 다른 프로그래밍 언어의 함수와도 유사하다.  벡터를 생성하지만 특별한 벡터 또는 행렬 연산을 사용하지 않고 있다.  그러한 연산은 곧 볼 수 있을 것이다.

아래는 또 다른 피보나치 함수이다.  그 반환값은 단순히 ``n`` 번째 피보나치 수 이다.

In [3]:
def fibnum(n):
    """
    Fibonacci number
    generates the nth Fibonacci number.
    :param n: 
    :return: 
    """
    if 1 >= n:
        f = 1
    else:
        f = fibnum(n - 1) + fibnum(n - 2)
        
    return f


fibnum(12)

233

``fibnum`` 함수는 *재귀적* 이다. 실은 *재귀적*이라는 용어는 수학과 전산학 모두에서 사용된다. 관계식 $f_n = f_{n-1} +f_{n-2}$ 는 *점화식* 또는 *재귀식* 이라고 알려져 있고, (전산학에서) 스스로 자신을 호출하는 함수는 *재귀 함수*이다.

재귀적 프로그램은 멋있어 보이지만 댓가가 따른다. 실행 시간을 측정해 보자.

In [4]:
%%time
fibnum(24)

Wall time: 36 ms


75025

아래 명령은 실행하지 말기 바란다.

``%%time``

``fibnum(50)``

이제 ``goldfract(6)``과 ``fibonacci(7)``의 결과를 비교해 보자.  전자는 분수 21/13이고 후자는 13 과 21 로 끝난다.  이것은 우연이 아니다. 연분수는 아래 연산을 반복하는데

``p = p + q``

피보나치 수를 생성하는 식도

``f[k] = f[k - 1] + f[k - 2]``

이다.  사실 $\phi_n$ 으로 *n* 항 연분수 황금률을 표시하면 다음과 같다.

$$\frac{f_{n+1}}{f_n} = \phi_n$$

무한 극한에서, 연속되는 피보나치수의 비는 황금률에 수렴한다.

$$ \lim_{x\to\infty} \frac{f_{n+1}}{f_n}=\phi$$

이를 보기 위해 40개의 피보나치 수를 계산한다.

In [2]:
n = 40
f = fibonacci(n)

In [5]:
np.set_printoptions(15)
f[1:] / f[:n-1]

array([[ 2.               ],
       [ 1.5              ],
       [ 1.666666666666667],
       [ 1.6              ],
       [ 1.625            ],
       [ 1.615384615384615],
       [ 1.619047619047619],
       [ 1.617647058823529],
       [ 1.618181818181818],
       [ 1.617977528089888],
       [ 1.618055555555556],
       [ 1.618025751072961],
       [ 1.618037135278515],
       [ 1.618032786885246],
       [ 1.618034447821682],
       [ 1.618033813400125],
       [ 1.618034055727554],
       [ 1.618033963166706],
       [ 1.618033998521803],
       [ 1.618033985017358],
       [ 1.618033990175597],
       [ 1.618033988205325],
       [ 1.618033988957902],
       [ 1.618033988670443],
       [ 1.618033988780243],
       [ 1.618033988738303],
       [ 1.618033988754322],
       [ 1.618033988748204],
       [ 1.618033988750541],
       [ 1.618033988749648],
       [ 1.618033988749989],
       [ 1.618033988749859],
       [ 1.618033988749909],
       [ 1.61803398874989 ],
       [ 1.618

이것은 ``f[1]`` (``f``의 두번째 항) 부터 ``f[n-1]`` (``f``의 마지막 항) 까지의 벡터를, 각 요소별로, ``f[0]`` (``f``의 첫번째 항) 부터 ``f[n-2]`` (``f``의 마지막에서 두번째 항) 까지의 벡터로 나누는 것이다. ``n``을 40으로 택한 까닭을 알겠는가? 아래와 같이 황금률과 비교해 볼 수 있다.

In [6]:
phi = (1.0 + 5**0.5) * 0.5
f[1:] / f[:n-1] - phi

array([[  3.819660112501051e-01],
       [ -1.180339887498949e-01],
       [  4.863267791677184e-02],
       [ -1.803398874989481e-02],
       [  6.966011250105097e-03],
       [ -2.649373365279484e-03],
       [  1.013630297724166e-03],
       [ -3.869299263654646e-04],
       [  1.478294319232631e-04],
       [ -5.646066000730698e-05],
       [  2.156680566067770e-05],
       [ -8.237676933475768e-06],
       [  3.146528619657474e-06],
       [ -1.201864648914253e-06],
       [  4.590717870289751e-07],
       [ -1.753497695933248e-07],
       [  6.697765919660981e-08],
       [ -2.558318845657936e-08],
       [  9.771908393574336e-09],
       [ -3.732536946188247e-09],
       [  1.425702222945802e-09],
       [ -5.445699446937624e-10],
       [  2.080071670462758e-10],
       [ -7.945177848966978e-11],
       [  3.034772433352373e-11],
       [ -1.159183860011126e-11],
       [  4.427569422205124e-12],
       [ -1.691313755713963e-12],
       [  6.459277557269161e-13],
       [ -2.46

마지막 값은 얼마인가? 

피보나치의 토끼 우리의 마리수는 매달 두배로 늘어나는 것은 아니다. 매달 황금비율 만큼 늘어난다.

피보나치 수에 관한 닫힌 형태의 해를 구하는 것도 가능하다.  요점은 어떤 상수 *c* 와 $\rho$ 값에 대해 다음과 같은 형태의 해를 찾는 것이다.

$$f_n = c \rho^n$$

재귀식

$$f_n = f_{n-1} +f_{n-2}$$

은 

$$\rho^2 = \rho + 1$$

이 된다.  이 식은 앞에서 본 적이 있다.  가능한 $\rho$ 값은 두개로 즉 $\phi$ 와 $1 - \rho$ 이다. 점화식에 관한 일반해는 다음과 같다

$$f_n = c_1 \phi ^ n + c_2 (1-\phi)^n $$

상수 $c_1$ 과 $c_2$는 초기 조건에 의해 정해지는데 다음과 같이 쓸 수 있다.

\begin{align}
f_0 = c_1 + c_2 = 1 \\\
f_1 = c_1 \phi + c_2 (1 - \phi) = 1
\end{align}

행렬 계산 명령으로 풀 수도 있지만, 손으로 푸는 것이 더 쉽다.

\begin{align}
c_1 = \frac{\phi}{2\phi - 1} \\\
c_2 = -\frac{(1 - \phi)}{2\phi - 1}
\end{align}



In [7]:
import sympy as sy
c1, c2, phi = sy.symbols('c1, c2, phi')
eq1 = sy.Eq(c1 + c2, 1)
eq2 = sy.Eq(c1 * phi + c2 * (1 - phi), 1)
sy.solve((eq1, eq2), (c1, c2))

{c1: phi/(2*phi - 1), c2: (phi - 1)/(2*phi - 1)}

$$f_n = \frac{1}{2\phi - 1}(\phi^{n+1}-(1-\phi)^{n+1})$$